File: basketball-csharp/BasketballTickets/Repository/Impl/CustomerRepositoryImpl.cs
using BasketballTickets.Domain;
using BasketballTickets.Repository.Abstracts;
using BasketballTickets.Repository.Interfaces;
using BasketballTickets.Utils;
using Microsoft.Data.Sqlite;

namespace BasketballTickets.Repository.Impl;

public class CustomerRepositoryImpl : AbstractDBRepository<long, Customer>, ICustomerRepository
{
    public CustomerRepositoryImpl(DatabaseConfig dbConfig) : base(dbConfig) { }

    protected override string GetTableName() => "customers";

    protected override string GetInsertSql() =>
        "INSERT INTO customers (name, address) VALUES (@name, @address)";

    protected override string GetUpdateSql() =>
        "UPDATE customers SET name=@name, address=@address WHERE id=@id";

    protected override Customer ExtractEntity(SqliteDataReader r) =>
        new(r.GetInt64(r.GetOrdinal("id")),
            r.GetString(r.GetOrdinal("name")),
            r.GetString(r.GetOrdinal("address")));

    protected override void BindInsert(SqliteCommand cmd, Customer c)
    {
        cmd.Parameters.AddWithValue("@name", c.Name);
        cmd.Parameters.AddWithValue("@address", c.Address);
    }

    protected override void BindUpdate(SqliteCommand cmd, Customer c)
    {
        cmd.Parameters.AddWithValue("@name", c.Name);
        cmd.Parameters.AddWithValue("@address", c.Address);
        cmd.Parameters.AddWithValue("@id", c.Id);
    }

    protected override void SetGeneratedId(Customer entity, long id) => entity.Id = id;

    // ---- Metode specifice domeniului ----

    public IList<Customer> FindByName(string name)
    {
        Logger.Debug($"FindByName('{name}')");
        const string sql = "SELECT * FROM customers WHERE LOWER(name) = LOWER(@name)";
        return QueryWithParam(sql, ("@name", name));
    }

    public IList<Customer> FindByAddress(string address)
    {
        Logger.Debug($"FindByAddress('{address}')");
        const string sql = "SELECT * FROM customers WHERE LOWER(address) LIKE LOWER(@address)";
        return QueryWithParam(sql, ("@address", $"%{address}%"));
    }

    public IList<Customer> FindByNameOrAddress(string? name, string? address)
    {
        Logger.Debug($"FindByNameOrAddress(name='{name}', address='{address}')");

        var sql = "SELECT * FROM customers WHERE 1=1";
        var parameters = new List<(string, object)>();

        if (!string.IsNullOrWhiteSpace(name))
        {
            sql += " AND LOWER(name) = LOWER(@name)";
            parameters.Add(("@name", name));
        }
        if (!string.IsNullOrWhiteSpace(address))
        {
            sql += " AND LOWER(address) LIKE LOWER(@address)";
            parameters.Add(("@address", $"%{address}%"));
        }

        var results = new List<Customer>();
        using var conn = DbConfig.GetConnection();
        using var cmd = new SqliteCommand(sql, conn);
        foreach (var (paramName, value) in parameters)
            cmd.Parameters.AddWithValue(paramName, value);

        using var reader = cmd.ExecuteReader();
        while (reader.Read()) results.Add(ExtractEntity(reader));

        Logger.Debug($"FindByNameOrAddress → {results.Count} results");
        return results;
    }

    // ---- Helper ----
    private List<Customer> QueryWithParam(string sql, (string name, object value) param)
    {
        var results = new List<Customer>();
        using var conn = DbConfig.GetConnection();
        using var cmd = new SqliteCommand(sql, conn);
        cmd.Parameters.AddWithValue(param.name, param.value);
        using var reader = cmd.ExecuteReader();
        while (reader.Read()) results.Add(ExtractEntity(reader));
        return results;
    }
}

File: java/src/main/java/org/example/repository/impl/CustomerRepositoryImpl.java
package org.example.repository.impl;

import org.example.domain.Customer;
import org.example.repository.abstracts.AbstractDBRepository;
import org.example.repository.interfaces.CustomerRepository;
import org.example.utils.DatabaseConfig;

import java.sql.*;
import java.util.ArrayList;
import java.util.List;

public class CustomerRepositoryImpl
        extends AbstractDBRepository<Long, Customer>
        implements CustomerRepository {

    public CustomerRepositoryImpl(DatabaseConfig dbConfig) {
        super(dbConfig);
    }

    @Override
    protected String getTableName() { return "customers"; }

    @Override
    protected String getInsertSql() {
        return "INSERT INTO customers (name, address) VALUES (?, ?)";
    }

    @Override
    protected String getUpdateSql() {
        return "UPDATE customers SET name=?, address=? WHERE id=?";
    }

    @Override
    protected Customer extractEntity(ResultSet rs) throws SQLException {
        return new Customer(
                rs.getLong("id"),
                rs.getString("name"),
                rs.getString("address")
        );
    }

    @Override
    protected void bindInsert(PreparedStatement ps, Customer c) throws SQLException {
        ps.setString(1, c.getName());
        ps.setString(2, c.getAddress());
    }

    @Override
    protected void bindUpdate(PreparedStatement ps, Customer c) throws SQLException {
        ps.setString(1, c.getName());
        ps.setString(2, c.getAddress());
        ps.setLong(3, c.getId());
    }

    @Override
    protected void setGeneratedId(Customer entity, long id) { entity.setId(id); }

    // ---- Domain-specific queries ----

    @Override
    public List<Customer> findByName(String name) {
        logger.debug("findByName('{}')", name);
        String sql = "SELECT * FROM customers WHERE LOWER(name) = LOWER(?)";
        return queryWithParam(sql, name);
    }

    @Override
    public List<Customer> findByAddress(String address) {
        logger.debug("findByAddress('{}')", address);
        String sql = "SELECT * FROM customers WHERE LOWER(address) LIKE LOWER(?)";
        return queryWithParam(sql, "%" + address + "%");
    }

    @Override
    public List<Customer> findByNameOrAddress(String name, String address) {
        logger.debug("findByNameOrAddress(name='{}', address='{}')", name, address);
        StringBuilder sql = new StringBuilder("SELECT * FROM customers WHERE 1=1");
        List<String> params = new ArrayList<>();

        if (name != null && !name.isBlank()) {
            sql.append(" AND LOWER(name) = LOWER(?)");
            params.add(name);
        }
        if (address != null && !address.isBlank()) {
            sql.append(" AND LOWER(address) LIKE LOWER(?)");
            params.add("%" + address + "%");
        }

        List<Customer> result = new ArrayList<>();
        try (Connection conn = dbConfig.getConnection();
             PreparedStatement ps = conn.prepareStatement(sql.toString())) {
            for (int i = 0; i < params.size(); i++) ps.setString(i + 1, params.get(i));
            ResultSet rs = ps.executeQuery();
            while (rs.next()) result.add(extractEntity(rs));
        } catch (SQLException e) {
            logger.error("findByNameOrAddress failed: {}", e.getMessage(), e);
        }
        logger.debug("findByNameOrAddress → {} results", result.size());
        return result;
    }

    // ---- Helper ----
    private List<Customer> queryWithParam(String sql, String param) {
        List<Customer> result = new ArrayList<>();
        try (Connection conn = dbConfig.getConnection();
             PreparedStatement ps = conn.prepareStatement(sql)) {
            ps.setString(1, param);
            ResultSet rs = ps.executeQuery();
            while (rs.next()) result.add(extractEntity(rs));
        } catch (SQLException e) {
            logger.error("queryWithParam failed: {}", e.getMessage(), e);
        }
        return result;
    }
}